client/main.go

```diff
@@ -380,40 +387,70 @@ func main() {
 		}()
 	}
 
-	var wg sync.WaitGroup
-	wg.Add(*concurrency)
-	shutdownChannels := make([]chan struct{}, *concurrency)
+	var mainWait sync.WaitGroup
+	mainWait.Add(*connections)
 
-	for i := int(0); i < *concurrency; i++ {
+	concurrency := *connections * *streams
+	shutdownChannels := make([]chan struct{}, concurrency)
+	for c := int(0); c < concurrency; c++ {
 		shutdownChannel := make(chan struct{}, 2)
 		shutdownChannels = append(shutdownChannels, shutdownChannel)
+	}
+
+	var connOpts []grpc.DialOption
+	if *tlsTrustChainFile != "" {
+		creds, err := credentials.NewClientTLSFromFile(*tlsTrustChainFile, "")
+		if err != nil {
+			exUsage("invalid ca cert file")
+		}
+		connOpts = append(connOpts, grpc.WithTransportCredentials(creds))
+	} else {
+		connOpts = append(connOpts, grpc.WithInsecure())
+	}
 
+	for c := int(0); c < *connections; c++ {
 		go func() {
-			r := rand.New(rand.NewSource(time.Now().UnixNano()))
+			defer mainWait.Done()
+
+			var connWait sync.WaitGroup
+			connWait.Add(*streams)
+
 			// Set up a connection to the server.
-			conn, err := grpc.Dial(*address, grpc.WithInsecure())
+			conn, err := grpc.Dial(*address, connOpts...)
 			if err != nil {
 				log.Fatalf("did not connect: %v", err)
+				return
 			}
 			defer conn.Close()
 			client := pb.NewResponderClient(conn)
 
-			if !*streaming {
-				sendNonStreamingRequests(client, shutdownChannel, *clientTimeout,
-					lengthDistribution, latencyDistribution, float32(*errorRate), r, received)
-			} else {
-				err := sendStreamingRequests(client, shutdownChannel,
-					lengthDistribution, latencyDistribution, *streamingRatio, r, received)
-				if err != nil {
-					log.Fatalf("could not send a request: %v", err)
-				}
+			offset := *streams * c
+			for s := int(0); s < *streams; s++ {
+				shutdownChannel := shutdownChannels[offset+s]
```

Comment:
when the concurrency is higher than 1, it looks like we'll miss the first shutdown channels < `*streams * c` since that's the offset.